### DISCUSSION TITLE
Cap on `retryStrategy` backoff?

### DISCUSSION THREAD

**Discussion by eiriklid** [argoproj/argo-workflows#13584] [Q&A] [answered]
Is it possible to cap the backoff in `retryStrategy`?

I want to have quick retries in the beginning, but not have to have a exponential backoff as well.
The [Pod Backoff Failure Policy](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-backoff-failure-policy) from Kubernetes have been ideal:

> ...  Job are recreated by the Job controller with an exponential back-off delay (10s, 20s, 40s ...) **capped at six minutes.**

**Comment by agilgur5** [comment]
That's what the `maxDuration` field is for, per [the backoff example](https://raw.githubusercontent.com/argoproj/argo-workflows/release-3.5/examples/retry-backoff.yaml).

Argo's backoff also [inherits from k8s `apimachinery`'s `wait` package](https://github.com/argoproj/argo-workflows/blob/1b0f00803d513d10b0f6f1bc73e99c8d2440c576/util/wait/backoff.go#L13)

**Reply by eiriklid** [reply]
Not according to the [documentation](https://argo-workflows.readthedocs.io/en/latest/fields/#backoff).

> MaxDuration is the maximum amount of time allowed for a workflow in the backoff strategy. It is important to note that if the workflow template includes activeDeadlineSeconds, the pod's deadline is initially set with activeDeadlineSeconds. However, when the workflow fails, the pod's deadline is then overridden by maxDuration. This ensures that the workflow does not exceed the specified maximum duration when retries are involved.

When I set 
```yaml
retryStrategy:
      when: "Failed"
      limit: 100
      backoff:
        duration: 10s
        factor: 2
        maxDuration: 6m
```

The workflow is killed after 6 minutes, while with a Kubernetes Job/Cronjob it would retry after 10, 20, 40, 80, 160, 320, 360, 360, ... until it hits the limit or activeDeadlineSeconds.

**Reply by agilgur5** [reply, marked as answer]
> Not according to the [documentation](https://argo-workflows.readthedocs.io/en/latest/fields/#backoff).

Yea, my bad, you're correct. I keep mixing up what `maxDuration` means and that specific doc was recently updated actually: https://github.com/argoproj/argo-workflows/pull/13068.
Now that I'm remembering, I called out this specific field as super unclear in feedback when I beta tested the [CNCF Argo Certification](https://www.cncf.io/training/certification/capa/) back in January 😅 (there was a question specifically on it and I advocated to remove that question)

> Argo's backoff also [inherits from k8s `apimachinery`'s `wait` package](https://github.com/argoproj/argo-workflows/blob/1b0f00803d513d10b0f6f1bc73e99c8d2440c576/util/wait/backoff.go#L13)

In another layer of confusion, the [Argo `Backoff` spec](https://github.com/argoproj/argo-workflows/blob/dfcaca4b92e5d020b22aa65b7826ea8f37163b02/pkg/apis/workflow/v1alpha1/workflow_types.go#L2019) actually differs from [`apimachinery`'s `Backoff`](https://github.com/kubernetes/apimachinery/blob/acfda137309f7e4bb432290277d6c1ad4d554917/pkg/util/wait/backoff.go#L30) apparently. In `apimachinery`, `Backoff.Cap` sounds like what you're looking for, _literally_.
Looking at the `blame`s, it's possible that `Cap` didn't exist at the time Argo's `Backoff` was created: https://github.com/kubernetes/apimachinery/commit/e52d7e07dd281a5895a88fcd785cd52899ce72b3 / https://github.com/kubernetes/kubernetes/pull/71088 was a year before https://github.com/argoproj/argo-workflows/pull/1782, but using an older k8s version might've prevented usage. 

Would you like to file a feature request for this? Can reference this discussion and the above
**EDIT**: Feature request was filed in #13772

**Reply by eiriklid** [reply]
I would very much like a feature request for this! :) Thank you for your thorough investigation on the issue. How do we proceed with creating the feature request?

**Reply by Joibel** [reply]
Raise an issue at https://github.com/argoproj/argo-workflows/issues by choosing `New Issue` and then `Feature`. Please ensure you refer to this discussion, which you can do by putting the text `#13584` in the issue.

**Reply by agilgur5** [reply]
Feature request was filed in #13772
